### PR TITLE
[TRY] perf: spawn (parallelize) io tasks related to tarball

### PR DESCRIPTION
**Pull request by kdy1**
This PR uses the threadpool of `rayon` to perform actual blocking IO operations. It's done by using `rayon::scope` API.

image

Blocked by:
 - https://github.com/pnpm/pacquet/pull/188
 - https://github.com/pnpm/pacquet/pull/189

### CODE REVIEW

**Comment by kdy1** [comment]
Oops my commits are gone. I'll use `git reflog` to find it..

File: crates/tarball/Cargo.toml [hunk 1]
```diff
@@ -28,6 +28,7 @@ tar          = { workspace = true }
 tokio        = { workspace = true }
 zune-inflate = { workspace = true }
 tracing      = { workspace = true }
+rayon.workspace = true
```
Comment (KSXGitHub) [inline crates/tarball/Cargo.toml:31]:
```suggestion
rayon = { workspace = true }
```

(remember to align the columns, or just run `taplo format`)

File: crates/tarball/src/lib.rs [hunk 2]
```diff
@@ -201,73 +201,88 @@ impl<'a> DownloadTarballToStore<'a> {
...
+                    // Read the contents of the entry
+                    let mut buffer = Vec::with_capacity(entry.size() as usize);
+                    entry.read_to_end(&mut buffer).unwrap();
+
+                    let entry_path = entry.path().unwrap();
+                    let cleaned_entry_path =
+                        entry_path.components().skip(1).collect::<PathBuf>().into_os_string();
+                    let (file_path, file_hash) = store_dir
+                        .write_cas_file(scope, buffer, file_is_executable)
+                        .map_err(TarballError::WriteCasFile)?;
```
Comment (KSXGitHub) [inline crates/tarball/src/lib.rs:243]: This seems to be the only IO operation in the entire scope. And none of the operations that follow depend on this IO operation being done either. Meaning, we could potentially defer this operation after building the `cas_paths` (i.e. do the IO on the created `cas_paths`). See if this suggestion can improve the code quality and performance.

**Review by KSXGitHub — COMMENTED** [review]
Thank you for your works.

I have a few change requests.

**Comment by KSXGitHub** [comment]
A new PR has been merged into `main`. It allows `pacquet` to work with a big lockfile. You may pull from `main` (or rebase) and benchmark with this lockfile: https://gist.github.com/KSXGitHub/ca8076a18b384b6cb6eea86d4ec7c675

**Comment by KSXGitHub** [comment]
@kdy1 `main` is faster when run `just integrated-benchmark -s frozen-lockfile main HEAD -V -D ./crates/testing-utils/src/fixtures/big/` on this branch:

benchmark result

**Comment by kdy1** [comment]
It's expected because you are using a beast machine and the lockfile is for a very small project.

The PR is about limiting the number of threads used for syscalls.

**Comment by KSXGitHub** [comment]
> the lockfile is for a very small project

I added `-D ./crates/testing-utils/src/fixtures/big/` which contains a big lockfile.

> The PR is about limiting the number of threads used for syscalls.

Is there a mechanism to scale the number of syscalls limit based on number of CPUs? Yesterday, there was a merged PR that scale the concurrent network requests limit based on number of CPUs.

Also, is the `perf:` in the title misleading?

**Comment by kdy1** [comment]
`rayon` has an issue with a beast machine. https://github.com/rayon-rs/rayon/issues/394

Can you try limiting the number of rayon threads by configuring `RAYON_NUM_THREADS`? It's documented at https://docs.rs/rayon/latest/rayon/struct.ThreadPoolBuilder.html#method.num_threads

File: crates/tarball/Cargo.toml [hunk 3]
```diff
@@ -29,7 +29,7 @@ tar          = { workspace = true }
 tokio        = { workspace = true }
 zune-inflate = { workspace = true }
 tracing      = { workspace = true }
-rayon.workspace = true
+rayon = { workspace = true }
```
Comment (KSXGitHub) [inline crates/tarball/Cargo.toml:32]: You didn't align the column. You can either do it manually or run `taplo format`.

**Comment by KSXGitHub** [comment]
@kdy1 Should the `RAYON_NUM_THREADS` be low or high? Since I/O are blocking operations, which won't be benefited from having many parallel threads, it would be counterproductive to hog all CPUs for themselves. So counterintuitively, the more CPUs the machine has, the slower this code will be? Am I correct? If this is the case, then would it be more appropriate to have a dedicated thread reserved for only I/O operations?

**Comment by kdy1** [comment]
It should be lower.

> So counterintuitively, the more CPUs the machine has, the slower this code will be

This is partially correct, but this is not related to IO but rather a bug of `rayon`.
But as it's about communication cost, it goes like




```
         .
      .    .  
   .          .
.                .


```

The performance graph will look like this because `n` is very small here.  The communication cost is `O(n^2)`, which is still small if the CPU core is in small range

**Comment by KSXGitHub** [comment]
@kdy1 Is `RAYON_NUM_THREADS=16 just integrated-benchmark -s frozen-lockfile main HEAD -V` the correct way to run it? `main` was always 3 times faster than `HEAD` regardless of the number.

**Comment by kdy1** [comment]
Hmmm... Yeah, it's correct. So we should move syscalls to `spawn_blocking`, not `rayon`

**Comment by KSXGitHub** [comment]
@kdy1 Actually, my hypothesis regarding I/O operations is proven incorrect.

I have created a [test repo](https://github.com/KSXGitHub/test-io-parallel-theory) which benchmark 4 cases. The result was parallel I/O + serial CPU is the fastest.

**Pull request closed without merging**